9333 default observer fix

jlitzingerdev opened the pull request:
This is a fix for 9333: https://twistedmatrix.com/trac/ticket/9333

Some items of note for reviewers:

 1.  The new formatEventWithTracebacks API was deliberately kept private to avoid adding new public APIs to maintain. Additionally, its easy to make it public if so specified in review.

2.  The existing formatEventAsClassicLogText has similar code that can be replaced. Rather than assume that was desired, I kept the change as focused as possible. Easy to change if desired.

As noted in the ticket, there are cases where duplicate tracebacks result, specifically if observers are appended to the globalLogPublisher, rather than overwrite with a beginLoggingTo call on the globalLogBeginner. This is documented in the logger docs.

jlitzingerdev committed:
Add formatEventWithTraceback.

This API will format log events and append a traceback if the
log_failure key is present.  The existing formatEvent function cannot
be used without altering its semantics and breaking at least one
consumer (users (e.g. textFileLogObserver).

This API is intentionally unexported from twisted.logger.  If it has
value as a public API it can be exported later.

jlitzingerdev committed:
Ensure default observers always print tracebacks.

Currently, events from the new log system with a log_failure key do
not show tracebacks.  The formatEvent API does not append tracebacks,
so use the formatEventWithTraceback API to ensure tracebacks are printed.

jlitzingerdev committed:
Add newsfragment.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

hawkowl committed:
Merge branch 'trunk' into 9333-default-observer-fix

hawkowl commented:
Pushed to the buildbots.

hawkowl commented:
The OSX build failures are spurious. They pass in circlecli.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

markrwilliams commented on src/twisted/logger/_format.py:
```diff
@@ -287,3 +287,33 @@ def formatWithCall(formatString, mapping):
+    except KeyError:
+        return eventText
+    except:
+        traceback = u"(UNABLE TO OBTAIN TRACEBACK FROM EVENT)\n"
```
This should be lower case and contain the exception message, like [its counterpart in the legacy logging system](https://github.com/twisted/twisted/blob/104394ddb76775c57c18897160a5ad055617b51a/src/twisted/python/log.py#L447).

jlitzingerdev replied:
@markrwilliams As with the bare except, I modeled this after the existing similar example in `formatEventAsClassicLogText`, rather than the legacy log system.  Should `formatEventAsClassicLogText` be changed to match the legacy (and concatenate the error) in a separate ticket?

markrwilliams commented on src/twisted/logger/_format.py:
```diff
@@ -287,3 +287,33 @@ def formatWithCall(formatString, mapping):
+            traceback = traceback.decode('utf-8')
+    except KeyError:
+        return eventText
+    except:
```
This should be `except Exception:` - we don't want to catch `SystemExit`!

jlitzingerdev replied:
No problem.  I modeled formatEventWithTraceback after formatEventAsClassicLogText (above it), which also uses all caps and an bare except.  Is that something that should be fixed (in a different ticket/PR)?

markrwilliams replied:
Yes, I think `formatEventAsClassicLogText` should get changed via a new ticket.  I appreciate that you started with a goal of consistency, as well!

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+class FormatEventWithTracebackTests(unittest.TestCase):
+    """
+    Tests for L{formatEventWithTraceback}, all of which ensure that the
+    returned type is a UTF-8.
```
This is an incomplete sentence.  I think it should be, "... all of which ensure that the returned type is UTF-8 decoded text."

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        event["log_failure"] = f
+        eventText = formatEventWithTraceback(event)
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('Traceback', eventText)
```
A better test would be:

```python
try:
    raise CapturedError("This is a fake error")
except CapturedError:
    f = Failure()
...
self.assertIn(f.getTraceback().decode('utf-8'), eventText)
...
```

That's because it would exercise the code with a real traceback.

jlitzingerdev replied:
Makes sense.  What I submit will be slightly different as eventText is always unicode and getTraceback() is the native string type (the decode call fails on P3).

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        event["log_failure"] = f
+        eventText = formatEventWithTraceback(event)
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('Traceback', eventText)
```
As above.

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        event = {
+            "log_format": "{evil()}",
+            "evil": lambda: 1 / 0,
+            Unformattable(): "gurk",
```
This is extraneous to the stated purpose of the test, which is served by `{evil()}`.  Please remove it.

jlitzingerdev replied:
@markrwilliams I agree, with concerns.  It tests a valid "unformattable" path, but also neglects one.  The first path would be caught by your suggested change, eliminating the `Unformattable()`.  The second path, exercised by the `Unformattable()` key occurs when the __repr__ call for the event throws.  My question -- do we only need to exercise the former, or should I split this into two tests, one for each?  For the sake of completeness, that seems like the right thing to do, thoughts?

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        eventText = formatEventWithTraceback(event)
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('Traceback', eventText)
+        self.assertIn('This is a fake error', eventText)
```
This should be a unicode literal.

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('Traceback', eventText)
+        self.assertIn('This is a fake error', eventText)
+        self.assertIn('This is a test log message', eventText)
```
This should be a unicode literal.

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        eventText = formatEventWithTraceback(event)
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('Traceback', eventText)
+        self.assertIn('This is a fake error', eventText)
```
This should be a unicode literal.

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        event["log_failure"] = f
+        eventText = formatEventWithTraceback(event)
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('MESSAGE LOST', eventText)
```
This should be a unicode literal.

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        eventText = formatEventWithTraceback(event)
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('MESSAGE LOST', eventText)
+        self.assertIn('Traceback', eventText)
```
This should be a unicode literal.

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('MESSAGE LOST', eventText)
+        self.assertIn('Traceback', eventText)
+        self.assertIn('This is a fake error', eventText)
```
This should be a unicode literal.

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        event["log_failure"] = object()
+        eventText = formatEventWithTraceback(event)
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn("(UNABLE TO OBTAIN TRACEBACK FROM EVENT)", eventText)
```
This should be a unicode literal - also see above comment about the message in this case.

markrwilliams commented on src/twisted/logger/test/test_format.py:
```diff
@@ -429,3 +430,94 @@ class Unformattable(object):
+        }
+        eventText = formatEventWithTraceback(event)
+        self.assertIsInstance(eventText, unicode)
+        self.assertIn('This is a test log message', eventText)
```
This should be a unicode literal.

markrwilliams commented on src/twisted/logger/test/test_global.py:
```diff
@@ -356,3 +357,16 @@ def test_warningsModule(self):
+        log = Logger(observer=self.publisher)
+        log.failure('a failure', failure=f)
+        msg = self.errorStream.getvalue()
+        self.assertIn('a failure', msg)
```
This (and what follows) should _not_ be a unicode literal :)

That's because [FileLogObserver encodes unicode messages to UTF-8 when it's given a binary file](https://github.com/twisted/twisted/blob/104394ddb76775c57c18897160a5ad055617b51a/src/twisted/logger/_file.py#L33-L37), and [it is on Python 2](https://github.com/jlitzingerdev/twisted/blob/a0a1c7c2bd7b7a2f396ea6ddb0453c131ccec4fc/src/twisted/logger/test/test_global.py#L65).

jlitzingerdev replied:
@markrwilliams Am I correct in interpreting that this code should remain as-is, to exercise the different paths for a native string on Python2/Python3?  Or, did you want me to ensure the encoding was forced on both?

markrwilliams replied:
Sorry for being unclear!  This was more of a note to myself confirming that this code is correct to use a native string.

markrwilliams reviewed (approved):
This is an important improvement and a well-done PR.  Thank you so much!

Please address the review comments before resubmitting for review.

markrwilliams reviewed (changes requested):
This review is to block merging until comments are addressed.

markrwilliams committed:
Merge branch 'trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Eliminate bare except and catch Exception.

Caught during review, the code as-is would catch everything, including
SystemExit, which was not the intent.

jlitzingerdev committed:
Ensure literals are unicode.

Changes requested in review.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Refactor tests to capture and format an actual traceback.

Changes made as a result of review feedback.

jlitzingerdev committed:
Update class docstring.

Requested during review.

jlitzingerdev committed:
Refactor unformattable tests.

The review comment requested removal of the Unformattable() object in
the event key as it was extraneous to the test.  This is true, but it
does represent a valid test path on its own -- the path where the
object throws an exception when __repr__ is invoked.  Rather than
discard this case, I split the test into two tests.

jlitzingerdev committed:
Fix case/literal type of error message and append exception.

These changes were requested during review, they ensure the behavior
matches the legacy log system.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Address twistedchecker issues.

Two line length issues, and one missing blank line between class methods.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9333-default-observer-fix

jlitzingerdev commented on src/twisted/logger/_format.py:
```diff
@@ -287,3 +287,33 @@ def formatWithCall(formatString, mapping):
+    eventText = formatEvent(event)
+    try:
+        traceback = event["log_failure"].getTraceback()
+        if isinstance(traceback, bytes):
```
I have concerns with this code.  Failure's getTraceback() method will return a native string depending on the Python version, so str/bytes Python 2.  The assumption that the bytes can sanely decoded as utf-8 is not one I've been able to verify and is probably wrong.

Looking at `formatEventAsClassicLogText`, which I used as a guide, it doesn't do any conversion but seems (I may be misreading) to rely on the fact that bytes and unicode can be joined in Python2 (https://github.com/twisted/twisted/blob/2e7b403c03e9b9b891a5eb36297e494f541d5d6a/src/twisted/logger/_format.py#L192).  That may (haven't confirmed) expose it to the same issue as 9442 due to the ascii conversion.  Should I confirm that I'll open a ticket.

So while this might be fine, I'd like to do a bit more diligence before this is merged, thus, I'm removing it from review.

jlitzingerdev replied:
Related to the concern with `formatEventAsClassicLogText`:
https://twistedmatrix.com/trac/ticket/9456#ticket

jlitzingerdev commented:
Closing, no longer relevant.
